waku/v2/protocol/filter/filter_test.go

```diff
@@ -520,6 +645,408 @@ func (s *FilterTestSuite) TestAutoShard() {
+	}
+
+	// All messages should be received
+	s.waitForMessages(func() {
+		s.publishMessages(messages)
+	}, s.subDetails, messages)
+
+	_, err := s.lightNode.UnsubscribeAll(s.ctx)
+	s.Require().NoError(err)
+
+}
+
+func (s *FilterTestSuite) TestPubSubMultiOverlapContentTopic() {
+	var (
+		testTopic        = "/waku/2/go/filter/test"
+		testContentTopic = "Topic"
+		testPayload      = "test_msg"
+		suffix           string
+		messages         []WakuMsg
+	)
+
+	// Create test context
+	ctx, cancel := context.WithTimeout(context.Background(), 20*time.Second) // Test can't exceed 10 seconds
+	s.ctx = ctx
+	s.ctxCancel = cancel
+
+	// Prepare data
+	for i := 0; i < 10; i++ {
+		suffix = fmt.Sprintf("%02d", i)
+		messages = append(messages, WakuMsg{
+			pubSubTopic:  testTopic,
+			contentTopic: testContentTopic + suffix,
+			payload:      testPayload + suffix,
+		})
+	}
+
+	// Subscribe
+	for _, m := range messages {
+		s.subDetails = s.subscribe(m.pubSubTopic, m.contentTopic, s.fullNodeHost.ID())
+	}
+
+	// All messages should be received
+	s.waitForMessages(func() {
+		s.publishMessages(messages)
+	}, s.subDetails, messages)
+
+	_, err := s.lightNode.UnsubscribeAll(s.ctx)
+	s.Require().NoError(err)
+
+}
+
+func (s *FilterTestSuite) TestSubscriptionRefresh() {
+	var (
+		testPayload = "test_msg"
+		suffix      string
+		messages    []WakuMsg
+	)
+
+	// Prepare data
+	for i := 0; i < 2; i++ {
+		suffix = fmt.Sprintf("%02d", i)
+		messages = append(messages, WakuMsg{
+			pubSubTopic:  s.testTopic,
+			contentTopic: s.testContentTopic,
+			payload:      testPayload + suffix,
+		})
+	}
+
+	// Initial subscribe
+	s.subDetails = s.subscribe(s.testTopic, s.testContentTopic, s.fullNodeHost.ID())
+
+	// Repeat the same subscribe
+	s.subDetails = s.subscribe(s.testTopic, s.testContentTopic, s.fullNodeHost.ID())
+
+	// Both messages should be received
+	s.waitForMessages(func() {
+		s.publishMessages(messages)
+	}, s.subDetails, messages)
+
+	_, err := s.lightNode.UnsubscribeAll(s.ctx)
+	s.Require().NoError(err)
+
+}
+
+func (s *FilterTestSuite) TestContentTopicsLimit() {
+	var (
+		testTopic        = "/waku/2/go/filter/test"
+		testContentTopic = "Topic"
+		testPayload      = "test_msg"
+		suffix           string
+		maxContentTopics = 30
+		messages         []WakuMsg
+	)
+
+	// Create test context
+	ctx, cancel := context.WithTimeout(context.Background(), 20*time.Second) // Test can't exceed 10 seconds
+	s.ctx = ctx
+	s.ctxCancel = cancel
+
+	// Detect existing content topics from previous test
+	if len(s.contentFilter.PubsubTopic) > 0 {
+		existingTopics := len(s.contentFilter.ContentTopicsList())
+		if existingTopics > 0 {
+			maxContentTopics = maxContentTopics - existingTopics
+		}
+	}
+
+	// Prepare data
+	for i := 0; i < maxContentTopics; i++ {
+		suffix = fmt.Sprintf("%02d", i)
+		messages = append(messages, WakuMsg{
+			pubSubTopic:  testTopic,
+			contentTopic: testContentTopic + suffix,
+			payload:      testPayload + suffix,
+		})
```

Comment:
Actually, this code seems to repeat very frequently? perhaps it's good to extract it into a non exported function

Reply:
Refactored in b5715d6a29aeef36a5b29c0db4caf20c7fdcd93d